### ISSUE TITLE
variable type mismatch when trying compile

### ISSUE THREAD

**Issue by johnnv1** [pytorch/pytorch#110696]
### 🐛 Describe the bug

When updating the pytorch on kornia [we found a weird issue when trying to compile the GuidedBlur module](https://github.com/kornia/kornia/pull/2605/files/aacbb4c6bd3c4cc4d1e03a7edc237e1c6165ef6c#r1348152475). The error trace looks like it's some data type issue or related `aten.reflection_pad2d`, but I don't understand this issue. 

- The CI error log: https://github.com/kornia/kornia/actions/runs/6411290175/job/17408312107#step:5:21801
- The module we are trying to compile: https://github.com/kornia/kornia/blob/03e71e3009ed8a304f41a7ab2ab00da0975a9aab/kornia/filters/guided.py#L164

Locally, I also get it working for a group of parameters and not for the others:
Working with:
- kernel_size=5; subsample=1;

not working with:
- kernel_size=5; subsample=2;
- kernel_size=(5, 7); subsample=1;
- kernel_size=(5, 7); subsample=2;

The compile is working before for it, but I wasn't able to reduce it to a minimum reproducible example. I also didn't find related issues.

### Error logs

```output
venv/lib/python3.10/site-packages/torch/nn/modules/module.py:1518: in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/nn/modules/module.py:1527: in _call_impl
    return forward_call(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py:328: in _fn
    return fn(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/nn/modules/module.py:1518: in _wrapped_call_impl
    return self._call_impl(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/nn/modules/module.py:1527: in _call_impl
    return forward_call(*args, **kwargs)
kornia/filters/guided.py:213: in forward
    return guided_blur(guidance, input, self.kernel_size, self.eps, self.border_type, self.subsample)
kornia/filters/guided.py:161: in guided_blur
    return _guided_blur_multichannel_guidance(guidance, input, kernel_size, eps, border_type, subsample)
kornia/filters/guided.py:75: in _guided_blur_multichannel_guidance
    guidance_sub, input_sub, kernel_size = _preprocess_fast_guided_blur(guidance, input, kernel_size, subsample)
venv/lib/python3.10/site-packages/torch/_dynamo/eval_frame.py:490: in catch_errors
    return callback(frame, cache_entry, hooks, frame_state)
venv/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:641: in _convert_frame
    result = inner_convert(frame, cache_size, hooks, frame_state)
venv/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:133: in _fn
    return fn(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:389: in _convert_frame_assert
    return _compile(
venv/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:569: in _compile
    guarded_code = compile_inner(code, one_graph, hooks, transform)
venv/lib/python3.10/site-packages/torch/_dynamo/utils.py:189: in time_wrapper
    r = func(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:491: in compile_inner
    out_code = transform_code_object(code, transform)
venv/lib/python3.10/site-packages/torch/_dynamo/bytecode_transformation.py:1028: in transform_code_object
    transformations(instructions, code_options)
venv/lib/python3.10/site-packages/torch/_dynamo/convert_frame.py:458: in transform
    tracer.run()
venv/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:2074: in run
    super().run()
venv/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:724: in run
    and self.step()
venv/lib/python3.10/site-packages/torch/_dynamo/symbolic_convert.py:702: in step
    self.output.compile_subgraph(
venv/lib/python3.10/site-packages/torch/_dynamo/output_graph.py:857: in compile_subgraph
    self.compile_and_call_fx_graph(tx, pass2.graph_output_vars(), root)
venv/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
venv/lib/python3.10/site-packages/torch/_dynamo/output_graph.py:957: in compile_and_call_fx_graph
    compiled_fn = self.call_user_compiler(gm)
venv/lib/python3.10/site-packages/torch/_dynamo/utils.py:189: in time_wrapper
    r = func(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/_dynamo/output_graph.py:1024: in call_user_compiler
    raise BackendCompilerFailed(self.compiler_fn, e).with_traceback(
venv/lib/python3.10/site-packages/torch/_dynamo/output_graph.py:1009: in call_user_compiler
    compiled_fn = compiler_fn(gm, self.example_inputs())
venv/lib/python3.10/site-packages/torch/_dynamo/repro/after_dynamo.py:117: in debug_wrapper
    compiled_gm = compiler_fn(gm, example_inputs)
venv/lib/python3.10/site-packages/torch/__init__.py:1568: in __call__
    return compile_fx(model_, inputs_, config_patches=self.config)
venv/lib/python3.10/site-packages/torch/_inductor/compile_fx.py:1150: in compile_fx
    return aot_autograd(
venv/lib/python3.10/site-packages/torch/_dynamo/backends/common.py:55: in compiler_fn
    cg = aot_module_simplified(gm, example_inputs, **kwargs)
venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:3891: in aot_module_simplified
    compiled_fn = create_aot_dispatcher_function(
venv/lib/python3.10/site-packages/torch/_dynamo/utils.py:189: in time_wrapper
    r = func(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:3429: in create_aot_dispatcher_function
    compiled_fn = compiler_fn(flat_fn, fake_flat_args, aot_config, fw_metadata=fw_metadata)
venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:2212: in aot_wrapper_dedupe
    return compiler_fn(flat_fn, leaf_flat_args, aot_config, fw_metadata=fw_metadata)
venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:2392: in aot_wrapper_synthetic_base
    return compiler_fn(flat_fn, flat_args, aot_config, fw_metadata=fw_metadata)
venv/lib/python3.10/site-packages/torch/_functorch/aot_autograd.py:1573: in aot_dispatch_base
    compiled_fw = compiler(fw_module, flat_args)
venv/lib/python3.10/site-packages/torch/_dynamo/utils.py:189: in time_wrapper
    r = func(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/_inductor/compile_fx.py:1092: in fw_compiler_base
    return inner_compile(
venv/lib/python3.10/site-packages/torch/_dynamo/repro/after_aot.py:80: in debug_wrapper
    inner_compiled_fn = compiler_fn(gm, example_inputs)
venv/lib/python3.10/site-packages/torch/_inductor/debug.py:228: in inner
    return fn(*args, **kwargs)
venv/lib/python3.10/contextlib.py:79: in inner
    return func(*args, **kwds)
venv/lib/python3.10/site-packages/torch/_inductor/compile_fx.py:54: in newFunction
    return old_func(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/_inductor/compile_fx.py:341: in compile_fx_inner
    compiled_graph: CompiledFxGraph = fx_codegen_and_compile(
venv/lib/python3.10/site-packages/torch/_inductor/compile_fx.py:550: in fx_codegen_and_compile
    graph.run(*example_inputs)
venv/lib/python3.10/site-packages/torch/_dynamo/utils.py:189: in time_wrapper
    r = func(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/_inductor/graph.py:456: in run
    return super().run(*args)
venv/lib/python3.10/site-packages/torch/fx/interpreter.py:138: in run
    self.env[node] = self.run_node(node)
venv/lib/python3.10/site-packages/torch/_inductor/graph.py:722: in run_node
    result = super().run_node(n)
venv/lib/python3.10/site-packages/torch/fx/interpreter.py:195: in run_node
    return getattr(self, n.op)(n.target, args, kwargs)
venv/lib/python3.10/site-packages/torch/_inductor/graph.py:613: in call_function
    raise LoweringException(e, target, args, kwargs).with_traceback(
venv/lib/python3.10/site-packages/torch/_inductor/graph.py:610: in call_function
    out = lowerings[target](*args, **kwargs)
venv/lib/python3.10/site-packages/torch/_inductor/lowering.py:279: in wrapped
    out = decomp_fn(*args, **kwargs)
venv/lib/python3.10/site-packages/torch/_inductor/lowering.py:3115: in reflection_pad2d
    ranges=[*batch, sympy.Integer(h + top + bot), sympy.Integer(w + left + right)],
venv/lib/python3.10/site-packages/sympy/core/cache.py:70: in wrapper
    retval = cfunc(*args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

cls = <class 'sympy.core.numbers.Integer'>, i = s0 + 3

    @cacheit
    def __new__(cls, i):
        if isinstance(i, str):
            i = i.replace(' ', '')
        # whereas we cannot, in general, make a Rational from an
        # arbitrary expression, we can make an Integer unambiguously
        # (except when a non-integer expression happens to round to
        # an integer). So we proceed by taking int() of the input and
        # let the int routines determine whether the expression can
        # be made into an int or whether an error should be raised.
        try:
            ival = int(i)
        except TypeError:
>           raise TypeError(
                "Argument of Integer should be of numeric type, got %s." % i)
E           torch._dynamo.exc.BackendCompilerFailed: backend='inductor' raised:
E           LoweringException: TypeError: Argument of Integer should be of numeric type, got s0 + 3.
E             target: aten.reflection_pad2d.default
E             args[0]: TensorBox(
E               View(
E                 StorageBox(
E                   Pointwise(
E                     'cpu',
E                     torch.float32,
E                     def inner_fn(index):
E                         i0, i1, i2, i3, i4 = index
E                         tmp0 = ops.load(arg0_1, i4 + 4 * i3 + 16 * i2 + 48 * i0)
E                         tmp1 = ops.load(arg0_1, i4 + 4 * i3 + 16 * i1 + 48 * i0)
E                         tmp2 = tmp0 * tmp1
E                         return tmp2
E                     ,
E                     ranges=[2, 3, 3, 4, 4],
E                     origin_node=mul,
E                     origins={mul}
E                   )
E                 ),
E                 size=[2, 9, 4, 4],
E                 reindex=lambda i0, i1, i2, i3: [i0, ModularIndexing(i1, 3, 3), ModularIndexing(i1, 1, 3), i2, i3],
E                 origins={mul, view_3}
E               )
E             )
E             args[1]: [((s0 - 1)//2), s0 - ((s0 - 1)//2) - 1, 1, 1]
E           
E           
E           You can suppress this exception and fall back to eager by setting:
E               import torch._dynamo
E               torch._dynamo.config.suppress_errors = True

```

### Versions

python 3.10.13
torch 2.1.0 

cc @ezyang @msaroufim @wconstab @bdhirsh @anijain2305

**Comment by ezyang** [comment]
what if you set dynamic=False on your torch.compile?

**Comment by johnnv1** [comment]
> what if you set dynamic=False on your torch.compile?

This works! Too slow, but works! When we should set the `dynamic` parameter?

**Comment by ezyang** [comment]
This just confirms to me that it's a dynamic shapes problem.

The inductor minifier did not work, is that right? https://pytorch.org/docs/stable/torch.compiler_troubleshooting.html TORCHDYNAMO_REPRO_AFTER="aot"

If it didn't, if you can give me non-minified repro instructions I'll take a look after the LF slide deadline. If you post a debug log with `TORCH_LOGS=+dynamo,+aot,dynamic,+inductor` that'd be helpful too

**Comment by johnnv1** [comment]
the minifier is working, here is it:

```python
import torch
from torch import tensor, device
import torch.fx as fx
from torch._dynamo.testing import rand_strided
from math import inf
import torch._inductor.inductor_prims

import torch._dynamo.config
import torch._inductor.config
import torch._functorch.config





isolate_fails_code_str = None



# torch version: 2.1.0
# torch cuda version: 11.8
# torch git version: 7bcf7da3a268b435777fe87c7794c382f444e86d


# CUDA Info: 
# nvcc not found
# GPU Hardware Info: 
# NVIDIA GeForce RTX 3060 Ti : 1 


from torch.nn import *
class Repro(torch.nn.Module):
    def __init__(self):
        super().__init__()
        self.register_buffer('_tensor_constant0', tensor(1.4660e+13))
        self.register_buffer('_tensor_constant1', tensor(1.4660e+13))

    
    
    def forward(self, arg0_1, arg1_1):
        _tensor_constant0 = self._tensor_constant0
        full_default = torch.ops.aten.full.default([], 14660154687488.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
        expand = torch.ops.aten.expand.default(full_default, [1, 3, arg1_1]);  full_default = None
        slice_1 = torch.ops.aten.slice.Tensor(expand, 0, 0, 9223372036854775807);  expand = None
        unsqueeze = torch.ops.aten.unsqueeze.default(slice_1, 1);  slice_1 = None
        expand_1 = torch.ops.aten.expand.default(unsqueeze, [-1, 3, -1, -1]);  unsqueeze = None
        sub = arg1_1 - 1
        floordiv = sub // 2
        sub_1 = sub - floordiv;  sub = None
        reflection_pad2d = torch.ops.aten.reflection_pad2d.default(arg0_1, [floordiv, sub_1, 1, 1]);  floordiv = sub_1 = None
        view = torch.ops.aten.view.default(expand_1, [3, 1, 3, arg1_1]);  expand_1 = None
        sym_size = torch.ops.aten.sym_size(reflection_pad2d, 3)
        view_1 = torch.ops.aten.view.default(reflection_pad2d, [-1, 3, 6, sym_size]);  reflection_pad2d = sym_size = None
        convolution = torch.ops.aten.convolution.default(view_1, view, None, [1, 1], [0, 0], [1, 1], False, [0, 0], 3);  view_1 = view = None
        permute = torch.ops.aten.permute.default(convolution, [0, 2, 3, 1]);  convolution = None
        unsqueeze_1 = torch.ops.aten.unsqueeze.default(arg0_1, 1)
        unsqueeze_2 = torch.ops.aten.unsqueeze.default(arg0_1, 2);  arg0_1 = None
        mul = torch.ops.aten.mul.Tensor(unsqueeze_1, unsqueeze_2);  unsqueeze_1 = unsqueeze_2 = None
        view_3 = torch.ops.aten.view.default(mul, [2, 9, 4, 4]);  mul = None
        _tensor_constant1 = self._tensor_constant1
        full_default_1 = torch.ops.aten.full.default([], 14660154687488.0, dtype = torch.float32, layout = torch.strided, device = device(type='cpu'), pin_memory = False)
        expand_2 = torch.ops.aten.expand.default(full_default_1, [1, 3, arg1_1]);  full_default_1 = None
        slice_2 = torch.ops.aten.slice.Tensor(expand_2, 0, 0, 9223372036854775807);  expand_2 = None
        unsqueeze_3 = torch.ops.aten.unsqueeze.default(slice_2, 1);  slice_2 = None
        expand_3 = torch.ops.aten.expand.default(unsqueeze_3, [-1, 9, -1, -1]);  unsqueeze_3 = None
        sub_2 = arg1_1 - 1
        floordiv_1 = sub_2 // 2
        sub_3 = sub_2 - floordiv_1;  sub_2 = None
        reflection_pad2d_1 = torch.ops.aten.reflection_pad2d.default(view_3, [floordiv_1, sub_3, 1, 1]);  view_3 = floordiv_1 = sub_3 = None
        view_4 = torch.ops.aten.view.default(expand_3, [9, 1, 3, arg1_1]);  expand_3 = arg1_1 = None
        sym_size_1 = torch.ops.aten.sym_size(reflection_pad2d_1, 3)
        view_5 = torch.ops.aten.view.default(reflection_pad2d_1, [-1, 9, 6, sym_size_1]);  reflection_pad2d_1 = sym_size_1 = None
        convolution_1 = torch.ops.aten.convolution.default(view_5, view_4, None, [1, 1], [0, 0], [1, 1], False, [0, 0], 9);  view_5 = view_4 = None
        permute_1 = torch.ops.aten.permute.default(convolution_1, [0, 2, 3, 1]);  convolution_1 = None
        view_7 = torch.ops.aten.view.default(permute_1, [2, 4, 4, 3, 3]);  permute_1 = None
        unsqueeze_4 = torch.ops.aten.unsqueeze.default(permute, -2)
        unsqueeze_5 = torch.ops.aten.unsqueeze.default(permute, -1)
        mul_1 = torch.ops.aten.mul.Tensor(unsqueeze_4, unsqueeze_5);  unsqueeze_4 = unsqueeze_5 = None
        sub_4 = torch.ops.aten.sub.Tensor(view_7, mul_1);  view_7 = mul_1 = None
        return (permute, sub_4)
        
def load_args(reader):
    buf0 = reader.storage(None, 384)
    reader.tensor(buf0, (2, 3, 4, 4), is_leaf=True)  # arg0_1
    reader.symint(4)  # arg1_1
load_args._version = 0
mod = Repro()
if __name__ == '__main__':
    from torch._dynamo.repro.after_aot import run_repro
    run_repro(mod, load_args, accuracy=False, command='minify', save_dir='/tmp/kornia/torch_compile_debug/run_2023_10_08_18_15_23_052128-pid_10577/minifier/checkpoints', tracing_mode='symbolic', check_str=None)
```

What is weird is, that if we isolate the code used in the test, and call it just one time it works (but will fail if try to compile again within other parameters) :)

The failing test is:
https://github.com/kornia/kornia/blob/e9d6f55b59c4d85137b178b906347ddd34a30a2c/test/filters/test_guided.py#L93

We pass the `torch.compile` as the fixture `torch_optimizer` (https://github.com/kornia/kornia/blob/e9d6f55b59c4d85137b178b906347ddd34a30a2c/conftest.py#L68)

if it is called just one time it will work fine, if we add a second call within other parameters it will fail -- the order of the call didn't change anything

```python
import torch

from kornia.filters import GuidedBlur
from kornia.testing import assert_close

torch.set_float32_matmul_precision('high')

backend = 'inductor'
device = torch.device('cpu')
dtype = torch.float32
kernel_size = (5, 7)
subsample = 1

guide = torch.ones(2, 3, 8, 8, device=device, dtype=dtype)
inpt = torch.ones(2, 3, 8, 8, device=device, dtype=dtype)

op = GuidedBlur(kernel_size, 0.1, subsample=subsample)
# will work fine
op_optimized = torch.compile(op, backend=backend)
out_a = op_optimized(guide, inpt)
out_b = op(guide, inpt)

assert_close(out_a, out_b)

kernel_size = 5
subsample = 2

guide = torch.ones(2, 3, 8, 8, device=device, dtype=dtype)
inpt = torch.ones(2, 3, 8, 8, device=device, dtype=dtype)

op = GuidedBlur(kernel_size, 0.1, subsample=subsample)
# will fail if we already called it before
op_optimized = torch.compile(op, backend=backend) 
out_a = op_optimized(guide, inpt)
out_b = op(guide, inpt)

assert_close(out_a, out_b)
```

**Comment by ezyang** [comment]
Yeah, the minifier is a little dorky, it doesn't know to force things to be dynamic (which is why calling it again triggers the failure). Thanks for getting it.

**Comment by ezyang** [comment]
I tested https://github.com/pytorch/pytorch/pull/110988 on the GuidedBlur repro and it appears to work. I don't know if it recompiles too much though.

**Comment by johnnv1** [comment]
to confirm, this fix did not go out into 2.1.1?
